api/src/wfl/service/google/bigquery.clj

```diff
@@ -0,0 +1,144 @@
+
+   Parameters
+   ----------
+   project  - Google Cloud Project to list the BigQuery datasets in.
+   dataset  - Dataset of the tables/views to list.
+
+   Example
+   -------
+     (list-tables \"broad-jade-dev-data\" \"zerosnapshot\")"
+  [project dataset]
+  (-> (str/join "/" ["projects" project "datasets" dataset "tables"])
+      (bigquery-url)
+      (http/get {:headers (once/get-auth-header)})
+      json-body
+      :tables))
+
+(def list-views "Function alias for views." list-tables)
+
+(defn query-table-sync
+  "Query for a BigQuery TABLE within a Data Repo SNAPSHOT in
+   Google Cloud PROJECT synchronously. Using non-legacy query
+   SQL syntax. Note: table and views are interchangeable BigQuery
+   concepts here, so both of them work as `TABLE`.
+
+   At least `data custodian` permission on the snapshot is
+   required for running the query job.
+
+   Parameters
+   ----------
+   project  - Google Cloud Project to list the BigQuery datasets in.
+   snapshot - Data Repo Snapshot.
+   table    - BigQuery table/view name.
+
+   Example
+   -------
+     (query-table-sync \"broad-jade-dev-data\" \"zerosnapshot\" \"sample\")"
+  [project snapshot table]
+  (let [query (format "SELECT * FROM `%s.%s.%s`" project snapshot table)]
+    (-> (str/join "/" ["projects" project "queries"])
+        (bigquery-url)
+        (http/post {:headers (once/get-auth-header)
+                    :body    (json/write-str
+                              {:query query
+                               :use_legacy_sql false})})
+        json-body)))
+
+(def query-view-sync "Function alias for views." query-table-sync)
+
+(defn dump-table->tsv
+  "Dump a BigQuery TABLE/view into a tsv FILE that's supported by Terra.
+   Will dump the tsv contents to bytes if no FILE is provided.
+
+   The first row header must follow the format 'entity:{data_table}_id'.
+   For example, 'entity:sample_id' will upload the tsv data into a `sample`
+   table in the workspace (or create one if it does not exist). If the
+   table already contains a sample with that id, it will get overwritten.
+
+   Parameters
+   ----------
+   table            - BigQuery table/view name.
+   terra-data-table - The `table` name in the Terra workspace to import the TSV.
+   file             - [optional] TSV file name to dump.
+
+   Example
+   -------
+
+     (dump-table->tsv table \"datarepo_row\" \"dumped.tsv\")
+     (dump-table->tsv table \"datarepo_row\")"
+  ([table terra-data-table file]
+   (letfn [(parse-row [row] (map :v (:f row)))
+           (format-header-for-terra [header]
+             (cons (format "entity:%s_id" terra-data-table) (rest header)))]
+     (let [headers (map :name (get-in table [:schema :fields]))
+           rows (map parse-row (get-in table [:rows]))
+           contents (-> []
+                        (into [(format-header-for-terra headers)])
+                        (into rows))]
+       (with-open [writer (io/writer file)]
```

Comment:
Instead of using a temporary file, maybe you can use a string stream? 
https://clojuredocs.org/clojure.core/with-out-str

Reply:
So if no file name is provided to this function, it will dump to bytes by default (which can then be used by `terra/import-entities` directly as `slurp` takes both bytes and file(names). I'm not sure if we will use this function to dump a tsv file locally (mainly for dev and testing purposes), so I intend to keep this, what do you think?

Reply:
I don't understand why you need to write this to a file because we're just going to read the contents right after to send to a terra workspace.

Reply:
OK having read the doc I didn't understand that passing `nil` to `io/writer` was a valid thing to do. *shudder*.
https://clojuredocs.org/clojure.java.io/writer

Reply:
OK having read the doc I didn't understand that passing `nil` to `io/writer` was a valid thing to do. *shudder*.
https://clojuredocs.org/clojure.java.io/writer